console/collections/benches/merkle_tree.rs

```diff
@@ -39,62 +40,144 @@ macro_rules! generate_leaves {
+            .into_iter()
+            .map(|leaf| {
+                let index: usize = Uniform::rand(&mut rng);
+                (index % num_leaves, leaf)
+            })
+            .collect::<Vec<_>>();
 
-        let merkle_tree = Testnet3::merkle_tree_bhp::<DEPTH>(&leaves).unwrap();
-        let new_leaf = generate_leaves!(1, &mut rng);
+        for num_new_leaves in UPDATE_SIZES {
+            // Construct a Merkle tree with the specified number of leaves.
+            let merkle_tree = Testnet3::merkle_tree_bhp::<DEPTH>(&leaves[..*num_leaves]).unwrap();
 
-        c.bench_function(
-            &format!("MerkleTree::append (adding single leaf to a tree with {num_leaves} leaves)"),
-            move |b| {
-                b.iter_with_setup(
+            c.bench_function(&format!("MerkleTree/update/{num_leaves}/{num_new_leaves}"), |b| {
+                b.iter_batched(
                     || merkle_tree.clone(),
                     |mut merkle_tree| {
-                        merkle_tree.append(&new_leaf).unwrap();
+                        for (index, new_leaf) in updates.iter().take(*num_new_leaves) {
+                            merkle_tree.update(*index, new_leaf).unwrap();
+                        }
                     },
+                    BatchSize::SmallInput,
                 )
-            },
-        );
+            });
+        }
+    }
+}
 
-        for num_new_leaves in APPEND_SIZES {
-            let merkle_tree = Testnet3::merkle_tree_bhp::<DEPTH>(&leaves).unwrap();
-            let new_leaves = generate_leaves!(*num_new_leaves, &mut rng);
-
-            c.bench_function(
-                &format!("MerkleTree::append (adding {num_new_leaves} new leaves to a tree with {num_leaves} leaves)"),
-                move |b| {
-                    b.iter_with_setup(
-                        || merkle_tree.clone(),
-                        |mut merkle_tree| {
-                            merkle_tree.append(&new_leaves).unwrap();
-                        },
-                    )
-                },
-            );
+fn batch_update(c: &mut Criterion) {
+    let mut rng = TestRng::default();
+    // Accumulate leaves in a vector to avoid recomputing across iterations.
+    let leaves = generate_leaves!(*NUM_LEAVES.last().unwrap(), &mut rng);
+    // For each number of leaves to update, benchmark the update operation.
+    for num_leaves in NUM_LEAVES {
+        // Generate all of the updates that will be applied to the tree.
+        // Note that we generate 2 * MAX_UPDATE_SIZE leaves to increase the number of unique of updates.
+        let mut updates = generate_leaves!(2 * *UPDATE_SIZES.last().unwrap(), &mut rng)
+            .into_iter()
+            .map(|leaf| {
+                let index: usize = Uniform::rand(&mut rng);
+                (index % num_leaves, leaf)
+            })
+            .collect::<Vec<_>>();
+        updates.sort_by(|(a, _), (b, _)| a.cmp(b));
```

Comment:
```suggestion
        updates.sort_by_key(|(a, _)| a);
```

Reply:
haven't tested it, might require dereferencing; also, since it is reversed right below, [`cmp::Reverse`](https://doc.rust-lang.org/std/cmp/struct.Reverse.html) could be applied here instead

Reply:
Sound good! Using `cmp::Reverse` seems to complicate the logic since we want to keep the most recent update.

Reply:
We need it so that `updates` is in descending order of index with the last update made to that index.
`updates.sort_by(|(a, _), (b, _)| b.cmp(a));` results in descending order of index with the first update made to that index.
Is there a different sequence of operations that works? 
I'm having a tough time coming up with one

Reply:
ah, ok; in that case I'd just add a comment on this so that it's not missed later on